### PR TITLE
Riedl

### PR DESCRIPTION
**Pull request by jriedl**

### CODE REVIEW

File: module/VuFind/src/VuFind/Controller/Plugin/PDASubscriptions.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+use VuFind\Exception\LoginRequired as LoginRequiredException,
+    Zend\Mvc\Controller\Plugin\AbstractPlugin,
+    VuFind\Db\Row\User, VuFind\Record\Cache;
+
+/**
+ * Zend action helper to perform favorites-related actions
+ */
+class PDASubscriptions extends AbstractPlugin
+{
+
```
Comment (ruschein) [inline module/VuFind/src/VuFind/Controller/Plugin/PDASubscriptions.php]: Please delete the superfluous blank line.

File: module/VuFind/src/VuFind/Db/Row/PDASubscription.php [hunk 2]
```diff
@@ -0,0 +1,24 @@
+<?php
+namespace VuFind\Db\Row;
+
+/**
+ * Row Definition for session
+ *
+ * @category VuFind
+ * @package  Db_Row
+ * @author   Demian Katz <demian.katz@villanova.edu>
```
Comment (ruschein) [inline module/VuFind/src/VuFind/Db/Row/PDASubscription.php]: You are not Demian Katz.

File: module/VuFind/src/VuFind/Db/Table/PDASubscription.php [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+     * @param string              $sort  Field to use for sorting (may include 'desc'
+     * qualifier)
+     *
+     * @return void
+     */
+    public static function applySort($query, $sort)
+    {
+        // Apply sorting, if necessary:
+        $legalSorts = [
+            'book_title', 'book_title desc', 'book_author', 'book_author desc', 'book_year', '_year desc'
```
Comment (ruschein) [inline module/VuFind/src/VuFind/Db/Table/PDASubscription.php]: One of these is not like the others.

File: module/VuFind/src/VuFind/Search/Results/Factory.php [hunk 4]
```diff
@@ -93,4 +93,22 @@ public static function getSubscriptions(ServiceManager $sm)
...
+     */
+    public static function getPDASubscriptions(ServiceManager $sm)
+    {
+        $factory = new PluginFactory();
+        $obj = $factory->createServiceWithName($sm, 'pdasubscriptions', 'PDASubscriptions');
+        $init = new \ZfcRbac\Initializer\AuthorizationServiceInitializer();
+        $init->initialize($obj, $sm);
+        return $obj;
+    }
+
```
Comment (ruschein) [inline module/VuFind/src/VuFind/Search/Results/Factory.php]: Superflous blank line.

File: module/ixTheo/src/ixTheo/Controller/MyResearchController.php [hunk 5]
```diff
@@ -6,6 +6,71 @@
 
 class MyResearchController extends \VuFind\Controller\MyResearchController
 {
+
```
Comment (ruschein) [inline module/ixTheo/src/ixTheo/Controller/MyResearchController.php]: Please remove this line.

File: module/ixTheo/src/ixTheo/Controller/MyResearchController.php [hunk 6]
```diff
@@ -6,6 +6,71 @@
...
+            return $this->createViewModel(
+                ['params' => $results->getParams(), 'results' => $results]
+            );
+        } catch (ListPermissionException $e) {
+            if (!$this->getUser()) {
+                return $this->forceLogin();
+            }
+            throw $e;
+        }
+
```
Comment (ruschein) [inline module/ixTheo/src/ixTheo/Controller/MyResearchController.php]: Delete this, please.

File: module/ixTheo/src/ixTheo/Controller/RecordController.php [hunk 7]
```diff
@@ -58,4 +58,62 @@ function subscribeAction()
...
     }
+
+
+    function processPDASubscribe()
+    {
+        if (!($user = $this->getUser())) {
+            return $this->forceLogin();
+        }
+        $post = $this->getRequest()->getPost()->toArray();
+        var_dump($post);
```
Comment (ruschein) [inline module/ixTheo/src/ixTheo/Controller/RecordController.php]: Please remove this.

File: module/ixTheo/src/ixTheo/Controller/RecordController.php [hunk 8]
```diff
@@ -58,4 +58,62 @@ function subscribeAction()
...
+        $results = $this->loadRecord()->pdasubscribe($post, $user);
+
+        if ($results == null) {
+            return $this->createViewModel();
+        }
+
+        $this->flashMessenger()->addMessage("Success", 'success');
+        return $this->redirectToRecord();
+    }
+
```
Comment (ruschein) [inline module/ixTheo/src/ixTheo/Controller/RecordController.php]: Too many blank lines.

File: module/ixTheo/src/ixTheo/RecordDriver/SolrMarc.php [hunk 9]
```diff
@@ -215,6 +215,40 @@ public function unsubscribe($params, $user)
         return $table->unsubscribe($userId, $recordId);
     }
 
+    public function pdasubscribe($params, $user)
+    {
+
```
Comment (ruschein) [inline module/ixTheo/src/ixTheo/RecordDriver/SolrMarc.php]: Ditto.

File: module/ixTheo/src/ixTheo/RecordDriver/SolrMarc.php [hunk 10]
```diff
@@ -215,6 +215,40 @@ public function unsubscribe($params, $user)
...
+            throw new LoginRequiredException('You must be logged in first');
+        }
+
+        $table = $this->getDbTable('PDASubscription');
+        $recordId = $this->getUniqueId();
+        $userId = $user->id;
+
+        return $table->unsubscribe($userId, $recordId);
+    }
+
```
Comment (ruschein) [inline module/ixTheo/src/ixTheo/RecordDriver/SolrMarc.php]: Ditto.

File: themes/ixTheoTheme/templates/record/pdasubscribe.phtml [hunk 11]
```diff
@@ -0,0 +1,20 @@
+<?= $this->flashmessages() ?>
+<?php if (!empty($this->pdasubscription)) { ?>
+    <?=$this->render($this->infoText);?>
+    <h4><?=$this->bookDescription;?></h4>
+    <textarea style="resize: none" name="addressfield" autofocus="true"  form="edit-save-form" rows="5" cols="30" 
+              required="true" placeholder="<?=$this->transEsc('Please enter your address here');?>"></textarea>
```
Comment (ruschein) [inline themes/ixTheoTheme/templates/record/pdasubscribe.phtml]: Should we add a colon after the word "here"?

File: themes/ixTheoTheme/templates/record/pdasubscribe.phtml [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    <p/>
+    <form id="edit-save-form" class="form-horizontal" method="post"
+          action="<?= $this->recordLink()->getActionUrl($this->driver, 'PDASubscribe') ?>" name="pdasubscribeRecord">
+        <input type="hidden" name="action" value="pdasubscribe"/>
+        <input class="btn btn-primary" type="submit" value="<?= $this->transEsc('PDASubscribe') ?>"/>
+    </form>
+<?php } else { ?>
+    <p><?=$this->transEsc("Subscription exists for this document");?>.</p>
```
Comment (ruschein) [inline themes/ixTheoTheme/templates/record/pdasubscribe.phtml]: Subs... -> A subs...

File: themes/ixTheoTheme/templates/record/pdasubscribe.phtml [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+    <form id="edit-save-form" class="form-horizontal" method="post"
+          action="<?= $this->recordLink()->getActionUrl($this->driver, 'PDASubscribe') ?>" name="pdaunsubscribeRecord">
+        <input type="hidden" name="action" value="pdaunsubscribe"/>
```
Comment (ruschein) [inline themes/ixTheoTheme/templates/record/pdasubscribe.phtml]: Here and many other places: Please use underscores to separate names.  E.g. pdaunsubscribe -> pda_unsubscribe  It's far easier to parse for human readers.

File: themes/ixTheoTheme/templates/static/de/PDASubscriptionInfoText.phtml [hunk 14]
```diff
@@ -0,0 +1,5 @@
+<p>Mittels Patron Driven Acquisition (PDA) kann die Beschaffung von Büchern veranlasst werden, die bisher noch nicht für die Ausleihe zur Verfügung stehen</p>
+<p>Falls Sie im unteren Feld eine Versandadresse angeben, werden wir versuchen,
```
Comment (ruschein) [inline themes/ixTheoTheme/templates/static/de/PDASubscriptionInfoText.phtml]: Missing period at end of sentence.

File: themes/ixTheoTheme/templates/static/en/PDASubscriptionInfoText.phtml [hunk 15]
```diff
@@ -0,0 +1,5 @@
+<p>By Patron Driven Acquisition (PDA) you can initiate the purchase of books that have not yet been available for loan so far</p>
+<p>If you enter your address in the area below we will try to buy the book and t
```
Comment (ruschein) [inline themes/ixTheoTheme/templates/static/en/PDASubscriptionInfoText.phtml]: Missing "o".

File: themes/ixTheoTheme/templates/static/en/PDASubscriptionInfoText.phtml [hunk 16]
```diff
@@ -0,0 +1,5 @@
...
+send it to you
```
Comment (ruschein) [inline themes/ixTheoTheme/templates/static/en/PDASubscriptionInfoText.phtml]: Missing period at end of sentence.

File: module/TueLib/src/TueLib/Controller/PDAProxyController.php [hunk 17]
```diff
@@ -60,45 +58,48 @@ protected function isAvailableForILL($isbn)
 
         // We use a several way scheme
         // 1.) If we have a match with ILL -> OK
-        // 2.) If can determine an ILL region take it as heuristics
-        // 3.) Parse the fullrecord
+        // 2.) Parse the fullrecord
```
Comment (jriedl) [inline module/TueLib/src/TueLib/Controller/PDAProxyController.php]: fullrecord is the name of the BSZ json field holding the XML representation

File: themes/ixTheoTheme/templates/record/pdasubscribe.phtml (hunk 11 above)
Comment (jriedl) [inline themes/ixTheoTheme/templates/record/pdasubscribe.phtml]: This is the end of statement for the php-transEsc

File: themes/ixTheoTheme/templates/record/pdasubscribe.phtml (hunk 12 above)
Comment (jriedl) [inline themes/ixTheoTheme/templates/record/pdasubscribe.phtml]: This is a label. Now fixed in English language file

File: themes/ixTheoTheme/templates/record/pdasubscribe.phtml (hunk 13 above)
Comment (jriedl) [inline themes/ixTheoTheme/templates/record/pdasubscribe.phtml]: Underscores are not common for zf2 index targets (see module.config.php)

**Pull request merged**